File: chapter-05/exercise-26-months-to-pay-off-a-credit-card/MonthsToPayOffACreditCard.java
class MonthsToPayOffACreditCard {
  public static void main(String[] args) {
    var balance = getd("What is your balance? ");
    var apr = getd("What is the APR on the card (as a percent)? ") / 100.0;
    var payment = getd("What is the monthly payment you can make? ");

    var months = calculateMonthsUntilPaidOff(balance, apr, payment);

    System.out.printf("%nIt will take you %d months to pay off this card.%n", months);
  }

  private static double getd(String prompt) {
    var line = System.console().readLine(prompt);
    try {
      return Double.parseDouble(line);
    } catch (NumberFormatException e) {
      System.out.println("Enter a valid number.");
      return getd(prompt);
    }
  }

  private static int calculateMonthsUntilPaidOff(double b, double apr, double p) {
    // i is the daily rate (APR divided by 365)
    var i = apr / 365.0;

    // -1 / 30
    var operand = -1 / 30.0;

    // log(1 + b/p (1 - (1 + i)^30))
    var dividend = Math.log(1 + (b / p) * (1 - Math.pow(1 + i, 30)));

    // log(1 + i)
    var divisor = Math.log(1 + i);

    return (int) (operand * (dividend / divisor));
  }
}

File: exercise-26-months-to-pay-off-a-credit-card/months_to_pay_off_a_credit_card.py
import math

def calculate_months_until_paid_off(b, apr, p):
    i = (apr / 100) / 365
    return (-1 / 30) * \
             ((math.log(1 + (b / p) * (1 - ((1 + i) ** 30)))) \
                 / (math.log(1 + i)))

balance = float(input("What is your balance? "))
apr = float(input("What is the APR on the card (as a percent)? "))
monthly_payment = float(input("What is the monthly payment you can make? "))
months = calculate_months_until_paid_off(balance, apr, monthly_payment)
print(f"It will take you {math.ceil(months)} months to pay off this card.")

